src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj

```diff
@@ -15,21 +15,19 @@
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;; Definitions
 
-(def jruby-puppet-env
-  "The environment variables that should be passed to the Puppet JRuby interpreters.
-  We don't want them to read any ruby environment variables, like $GEM_HOME or
-  $RUBY_LIB or anything like that, so pass it an empty environment map - except -
-  Puppet needs HOME and PATH for facter resolution, so leave those."
-  (select-keys (System/getenv) ["HOME" "PATH"]))
-
 (def ruby-code-dir
-  "The name of the directory containing the ruby code in this project.
-  This directory lives under src/ruby/"
-  "puppet-server-lib")
+  "The name of the directory containing the ruby code in this project."
+  "src/ruby/puppet-server-lib")
```

Comment:
Well, after some digging I figured out how this works.  It's documented at a high level, but sufficient enough that I feel like I understand it, in the [LoadService](http://jruby.org/apidocs/org/jruby/runtime/load/LoadService.html) JRuby java doc.

Basically, JRuby will search each element of the classpath by taking the name being required, map it onto each element of Ruby's LOAD_PATH, then map it again onto each element of the classpath.  This seems like a ridiculous huge number of permutations, but hey it seems to work!

More concerning to me is this behavior seems to assume there will be relative directories located on the Ruby LOAD_PATH.  That seems terrible and just plain wrong, but it's also how upstream JRuby fundamentally behaves, so we're at least relying on intended behavior upstream, if not sound behavior.